nixos/display-managers: add option to disable starting graphical-session

MrFreezeex opened the pull request:
###### Description of changes

Add a new option to allow to disable starting the
graphical-session.target automatically. This might be useful to disable to let the window manager starting the session itself rather than letting the display manager doing it before the window manager is actually started. Setting this to false can fix issues with systemd services starting too early on Wayland.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
nixos/display-managers: add option to disable starting graphical-session

Add a new option to allow to disable starting the
graphical-session.target automatically. This might be useful to disable to
let the window manager starting the session itself rather than letting the
display manager doing it before the window manager is actually started.
Setting this to false can fix issues with systemd services starting too early
on Wayland.

Co-Authored-By: Nicolas Berbiche <nicolas@normie.dev>
Signed-off-by: Arthur Outhenin-Chalandre <arthur.outhenin-chalandre@proton.ch>

MrFreezeex commented:
If this new option is disabled this should fix this home-manager issue https://github.com/nix-community/home-manager/issues/3599 (as pointed by @berbiche in https://github.com/nix-community/home-manager/issues/3599#issuecomment-1387233002 which I am merely upstreaming his patch, thanks to him very much!).

MrFreezeex commented:
Hi @flokli, sorry to ping you that way but I don't think there is technically a maintainer for this specific file (IIUC) and you are show as suggested reviewer on Github :see_no_evil:. If you are able to review this at some point in time that would be awesome :heart: :pray:.

flokli commented:
Hey @MrFreezeex, I don't really have a strong opinion about this. I'm mostly using gnome these days, and their login screen seem to be able to start an X gnome-shell and a wayland gnome-shell just fine.

You might want to ping some of the maintainers of the sway package, or people tinkering with wayland in general (git history of wayland packages).

MrFreezeex commented:
Hi @primeos @Synthetica9, as sway package maintainers, would you be able to review PR (or any suggestions on other possible reviewers)? :pray:

TLATER commented:
Just a drive-by comment since I'm currently trying to untangle this myself, but this would make having both Wayland *and* Xorg sessions more difficult.

I feel like the whole xinit stack in NixOS may need a review from ground up to make it support Wayland sessions, rather than little patches for specific issues like this one, with a clear flow for which scripts are expected to be run at what time for each compositor.

MrFreezeex commented:
> Just a drive-by comment since I'm currently trying to untangle this myself, but this would make having both Wayland _and_ Xorg sessions more difficult.
> 
> I feel like the whole xinit stack in NixOS may need a review from ground up to make it support Wayland sessions, rather than little patches for specific issues like this one, with a clear flow for which scripts are expected to be run at what time for each compositor.

I don't have a lot of experience with this part of the code but I guess it should be possible to start the session inside whatever display manager wether it is xorg or Wayland based maybe? You would still have to do this across all your sessions though this is true.

While I agree that there is probably some larger refactor to do but maybe there would be some more short term options for making this works without having to patch nixpkgs?

TLATER commented:
> While I agree that there is probably some larger refactor to do maybe there would be some shorter term options for making this works without having to patch nixpkgs while a longer term option has been decided/developed?

Probably, I just have long-term concerns with this patch, it'd be a pity if this was the last word in wayland session support ;)

As long as there is a proper tracking ticket for the refactor somewhere, and there's some active work towards it, this patch shouldn't cause any issues, and will make running wayland much easier in the short term.

MrFreezeex commented:
> Sorry for never responding btw. Unfortunately I cannot review this as I've switched to Sway to avoid graphical login/display managers but I should've still responded. If there are no `meta.maintainers`, like in this case, I usually use `git log` to check who frequently made relevant changes (but in this case it doesn't look that good... :o). Might be worth a try to ask via Matrix and/or Discord but I'm not sure. I guess you could also look for similar PRs and issues to see who handled them. We'd really need maintainers for this module or a team for that area (`maintainers/team-list.nix`).

Hey @primeos, no worries! I will try to ask around... I'm not quite sure of what you are implying by "I've switched to Sway to avoid graphical login/display managers" though, I am on sway too and this problem occurs with basically any systemd service that have a target on graphical-session. Like many services from home-manager are like that (wlsunset for instance, waybar if you set `systemd.enable). This problem never occurs if you have no systemd service with that kind of target and you have everything in your sway config with some `exec`.

primeos commented:
@MrFreezeex I mean that I start Sway from a tty/VT (which is the recommended / only supported approach btw: https://github.com/swaywm/sway/#running; it's also the only approach that we officially support in Nixpkgs/NixOS) and not via a login/display manager that executes that `xsession-wrapper` script (so I basically (intentionally) "bypass" the whole script anyway; I never really liked that script much tbh).
Plus I don't use any systemd user units that depend on that target (then again there might be some on my system that NixOS installs by default that I didn't notice).

Anyway, therefore this change shouldn't affect my setup at all (and I cannot (properly) review it).
